akka-actor/src/main/scala-2.12/akka/util/ByteString.scala

```diff
@@ -500,11 +500,13 @@ object ByteString {
       if (0 <= idx && idx < length) {
         var pos = 0
         var seen = 0
-        while (idx >= seen + bytestrings(pos).length) {
-          seen += bytestrings(pos).length
+        var frag = bytestrings(pos)
```

Comment:
Maybe worth trying out to see if it is even cheaper:
As far as I can see we always start from 0 and increase pos by 1 at a time, could it be even cheaper to use a `bytestrings.iterator` instead to move through the segments?

Reply:
I actually did try it with an iterator which was better in the worse case, but was worse in the best case, presumably due to the allocation of the `Iterator`
```
    2.13 original
    ByteString_apply_Benchmark.bss_apply_best_case   thrpt    3  206854021.793 ±  81500220.451  ops/s
    ByteString_apply_Benchmark.bss_apply_worst_case  thrpt    3     237125.194 ±    128394.832  ops/s

    2.13 optimized
    ByteString_apply_Benchmark.bss_apply_best_case   thrpt    3  209266780.913 ±  6821134.296  ops/s
    ByteString_apply_Benchmark.bss_apply_worst_case  thrpt    3     430348.094 ±    24412.915  ops/s

    2.13 iterator  
    ByteString_apply_Benchmark.bss_apply_best_case   thrpt    3   87246293.621 ± 106638565.695  ops/s
    ByteString_apply_Benchmark.bss_apply_worst_case  thrpt    3     467251.458 ±     92451.108  ops/s
``

Reply:
Thanks for the details, let's go with this then :)